SPI 3 Refactoring

chrisdutz opened the pull request:
In this refactoring, we are replacing the existing Netty-based core into a set of lightweight sub-modules without requiring Netty at all and with as little third party dependencies as possible.

This will allow using our code-generation framework independently from building PLC4J drivers and allow porting PLC4X to other languages as we no longer rely on external libraries for which counterparts need to be integrated for new languages.

chrisdutz committed:
chore: Started defining the new buffer-api kicking off the SPI3 refactoring (At this state the project should not be buildable)

chrisdutz committed:
chore: Removed some accidentally committed files.

dependabot[bot] committed:
build(deps): bump org.sonarsource.scanner.maven:sonar-maven-plugin (#2381)

Bumps [org.sonarsource.scanner.maven:sonar-maven-plugin](https://github.com/SonarSource/sonar-scanner-maven) from 5.4.0.6343 to 5.5.0.6356.
- [Release notes](https://github.com/SonarSource/sonar-scanner-maven/releases)
- [Commits](https://github.com/SonarSource/sonar-scanner-maven/compare/5.4.0.6343...5.5.0.6356)

---
updated-dependencies:
- dependency-name: org.sonarsource.scanner.maven:sonar-maven-plugin
  dependency-version: 5.5.0.6356
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump org.apache.commons:commons-text from 1.14.0 to 1.15.0 (#2380)

Bumps [org.apache.commons:commons-text](https://github.com/apache/commons-text) from 1.14.0 to 1.15.0.
- [Changelog](https://github.com/apache/commons-text/blob/master/RELEASE-NOTES.txt)
- [Commits](https://github.com/apache/commons-text/compare/rel/commons-text-1.14.0...rel/commons-text-1.15.0)

---
updated-dependencies:
- dependency-name: org.apache.commons:commons-text
  dependency-version: 1.15.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump com.google.googlejavaformat:google-java-format (#2379)

Bumps [com.google.googlejavaformat:google-java-format](https://github.com/google/google-java-format) from 1.32.0 to 1.33.0.
- [Release notes](https://github.com/google/google-java-format/releases)
- [Commits](https://github.com/google/google-java-format/compare/v1.32.0...v1.33.0)

---
updated-dependencies:
- dependency-name: com.google.googlejavaformat:google-java-format
  dependency-version: 1.33.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump golang.org/x/net from 0.47.0 to 0.48.0 in /plc4go (#2386)

Bumps [golang.org/x/net](https://github.com/golang/net) from 0.47.0 to 0.48.0.
- [Commits](https://github.com/golang/net/compare/v0.47.0...v0.48.0)

---
updated-dependencies:
- dependency-name: golang.org/x/net
  dependency-version: 0.48.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump golang.org/x/tools from 0.39.0 to 0.40.0 in /plc4go (#2385)

Bumps [golang.org/x/tools](https://github.com/golang/tools) from 0.39.0 to 0.40.0.
- [Release notes](https://github.com/golang/tools/releases)
- [Commits](https://github.com/golang/tools/compare/v0.39.0...v0.40.0)

---
updated-dependencies:
- dependency-name: golang.org/x/tools
  dependency-version: 0.40.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

sruehl committed:
chore(build): use different sdk site

https://github.com/raydac/gosdk-wrapper-maven-plugin/issues/11

dependabot[bot] committed:
build(deps): bump com.igormaznitsa:gosdk-wrapper-maven-plugin (#2390)

Bumps [com.igormaznitsa:gosdk-wrapper-maven-plugin](https://github.com/raydac/gosdk-wrapper-maven-plugin) from 1.1.1 to 1.1.2.
- [Release notes](https://github.com/raydac/gosdk-wrapper-maven-plugin/releases)
- [Changelog](https://github.com/raydac/gosdk-wrapper-maven-plugin/blob/main/CHANGELOG.md)
- [Commits](https://github.com/raydac/gosdk-wrapper-maven-plugin/compare/1.1.1...1.1.2)

---
updated-dependencies:
- dependency-name: com.igormaznitsa:gosdk-wrapper-maven-plugin
  dependency-version: 1.1.2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump mockito.version from 5.20.0 to 5.21.0 (#2389)

Bumps `mockito.version` from 5.20.0 to 5.21.0.

Updates `org.mockito:mockito-core` from 5.20.0 to 5.21.0
- [Release notes](https://github.com/mockito/mockito/releases)
- [Commits](https://github.com/mockito/mockito/compare/v5.20.0...v5.21.0)

Updates `org.mockito:mockito-junit-jupiter` from 5.20.0 to 5.21.0
- [Release notes](https://github.com/mockito/mockito/releases)
- [Commits](https://github.com/mockito/mockito/compare/v5.20.0...v5.21.0)

---
updated-dependencies:
- dependency-name: org.mockito:mockito-core
  dependency-version: 5.21.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
- dependency-name: org.mockito:mockito-junit-jupiter
  dependency-version: 5.21.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump com.gradle:develocity-maven-extension (#2391)

Bumps com.gradle:develocity-maven-extension from 2.2.1 to 2.3.

---
updated-dependencies:
- dependency-name: com.gradle:develocity-maven-extension
  dependency-version: '2.3'
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

chrisdutz committed:
doc: Added documentation for a set of test-structures and initializtaiton values for testing.

chrisdutz committed:
feat: Implemented a first version of the byte-based buffers along with enough tests for the 80% coverage.

dependabot[bot] committed:
build(deps): bump org.apache.maven.plugins:maven-release-plugin (#2397)

Bumps [org.apache.maven.plugins:maven-release-plugin](https://github.com/apache/maven-release) from 3.3.0 to 3.3.1.
- [Release notes](https://github.com/apache/maven-release/releases)
- [Commits](https://github.com/apache/maven-release/compare/maven-release-3.3.0...maven-release-3.3.1)

---
updated-dependencies:
- dependency-name: org.apache.maven.plugins:maven-release-plugin
  dependency-version: 3.3.1
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump actions/upload-artifact from 5 to 6 (#2396)

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 5 to 6.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump io.vavr:vavr from 0.10.7 to 0.11.0 (#2395)

Bumps [io.vavr:vavr](https://github.com/vavr-io/vavr) from 0.10.7 to 0.11.0.
- [Release notes](https://github.com/vavr-io/vavr/releases)
- [Commits](https://github.com/vavr-io/vavr/compare/v0.10.7...v0.11.0)

---
updated-dependencies:
- dependency-name: io.vavr:vavr
  dependency-version: 0.11.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump actions/cache from 4 to 5 (#2393)

Bumps [actions/cache](https://github.com/actions/cache) from 4 to 5.
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump logback.version from 1.5.21 to 1.5.22 (#2392)

Bumps `logback.version` from 1.5.21 to 1.5.22.

Updates `ch.qos.logback:logback-classic` from 1.5.21 to 1.5.22
- [Release notes](https://github.com/qos-ch/logback/releases)
- [Commits](https://github.com/qos-ch/logback/compare/v_1.5.21...v_1.5.22)

Updates `ch.qos.logback:logback-core` from 1.5.21 to 1.5.22
- [Release notes](https://github.com/qos-ch/logback/releases)
- [Commits](https://github.com/qos-ch/logback/compare/v_1.5.21...v_1.5.22)

---
updated-dependencies:
- dependency-name: ch.qos.logback:logback-classic
  dependency-version: 1.5.22
  dependency-type: direct:production
  update-type: version-update:semver-patch
- dependency-name: ch.qos.logback:logback-core
  dependency-version: 1.5.22
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump org.testcontainers:testcontainers (#2400)

Bumps [org.testcontainers:testcontainers](https://github.com/testcontainers/testcontainers-java) from 2.0.2 to 2.0.3.
- [Release notes](https://github.com/testcontainers/testcontainers-java/releases)
- [Changelog](https://github.com/testcontainers/testcontainers-java/blob/main/CHANGELOG.md)
- [Commits](https://github.com/testcontainers/testcontainers-java/compare/2.0.2...2.0.3)

---
updated-dependencies:
- dependency-name: org.testcontainers:testcontainers
  dependency-version: 2.0.3
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump org.testcontainers:junit-jupiter (#2399)

Bumps [org.testcontainers:junit-jupiter](https://github.com/testcontainers/testcontainers-java) from 1.21.3 to 1.21.4.
- [Release notes](https://github.com/testcontainers/testcontainers-java/releases)
- [Changelog](https://github.com/testcontainers/testcontainers-java/blob/main/CHANGELOG.md)
- [Commits](https://github.com/testcontainers/testcontainers-java/compare/1.21.3...1.21.4)

---
updated-dependencies:
- dependency-name: org.testcontainers:junit-jupiter
  dependency-version: 1.21.4
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump org.apache.karaf.tooling:karaf-maven-plugin (#2398)

Bumps org.apache.karaf.tooling:karaf-maven-plugin from 4.4.8 to 4.4.9.

---
updated-dependencies:
- dependency-name: org.apache.karaf.tooling:karaf-maven-plugin
  dependency-version: 4.4.9
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump org.jetbrains.kotlin:kotlin-stdlib (#2402)

Bumps [org.jetbrains.kotlin:kotlin-stdlib](https://github.com/JetBrains/kotlin) from 2.2.21 to 2.3.0.
- [Release notes](https://github.com/JetBrains/kotlin/releases)
- [Changelog](https://github.com/JetBrains/kotlin/blob/master/ChangeLog.md)
- [Commits](https://github.com/JetBrains/kotlin/compare/v2.2.21...v2.3.0)

---
updated-dependencies:
- dependency-name: org.jetbrains.kotlin:kotlin-stdlib
  dependency-version: 2.3.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump org.apache.maven:maven-artifact from 3.9.11 to 3.9.12 (#2401)

Bumps org.apache.maven:maven-artifact from 3.9.11 to 3.9.12.

---
updated-dependencies:
- dependency-name: org.apache.maven:maven-artifact
  dependency-version: 3.9.12
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

chrisdutz committed:
feat: Implemented a first version of the xml-based and ported the ascii-box-based buffers along with enough tests for the 80% coverage.

chrisdutz committed:
refactor: Updated the box-based buffer to operate without external dependencies.

chrisdutz committed:
refactor: Updated the box-based buffer to operate without external dependencies.

chrisdutz committed:
refactor: Updated the box-based buffer to operate without external dependencies.

chrisdutz committed:
refactor: Added the queryName to the browse request interceptor interface.

dependabot[bot] committed:
build(deps): bump dorny/test-reporter from 2.3.0 to 2.5.0 (#2409)

Bumps [dorny/test-reporter](https://github.com/dorny/test-reporter) from 2.3.0 to 2.5.0.
- [Release notes](https://github.com/dorny/test-reporter/releases)
- [Changelog](https://github.com/dorny/test-reporter/blob/main/CHANGELOG.md)
- [Commits](https://github.com/dorny/test-reporter/compare/fe45e9537387dac839af0d33ba56eed8e24189e8...b082adf0eced0765477756c2a610396589b8c637)

---
updated-dependencies:
- dependency-name: dorny/test-reporter
  dependency-version: 2.5.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

sruehl commented on plc4j/spi/buffers/ascii-box/src/main/java/org/apache/plc4x/java/spi/buffers/asciiboxbased/utils/ascii/AsciiBoxWriter.java:
```diff
@@ -235,16 +229,12 @@ public AsciiBox boxBelowBox(AsciiBox box1, AsciiBox box2) {
-            default:
-                return boxSideBySide(boxes.get(0), mergeHorizontal(new ArrayList<>(boxes).subList(1, boxes.size())));
-        }
+        return switch (boxes.size()) {
```
really like the new syntax... :)

dependabot[bot] committed:
build(deps): bump org.codehaus.mojo:exec-maven-plugin (#2405)

Bumps [org.codehaus.mojo:exec-maven-plugin](https://github.com/mojohaus/exec-maven-plugin) from 3.6.2 to 3.6.3.
- [Release notes](https://github.com/mojohaus/exec-maven-plugin/releases)
- [Commits](https://github.com/mojohaus/exec-maven-plugin/compare/3.6.2...3.6.3)

---
updated-dependencies:
- dependency-name: org.codehaus.mojo:exec-maven-plugin
  dependency-version: 3.6.3
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump net.bytebuddy:byte-buddy from 1.18.2 to 1.18.3 (#2404)

Bumps [net.bytebuddy:byte-buddy](https://github.com/raphw/byte-buddy) from 1.18.2 to 1.18.3.
- [Release notes](https://github.com/raphw/byte-buddy/releases)
- [Changelog](https://github.com/raphw/byte-buddy/blob/master/release-notes.md)
- [Commits](https://github.com/raphw/byte-buddy/compare/byte-buddy-1.18.2...byte-buddy-1.18.3)

---
updated-dependencies:
- dependency-name: net.bytebuddy:byte-buddy
  dependency-version: 1.18.3
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump logback.version from 1.5.22 to 1.5.23 (#2403)

Bumps `logback.version` from 1.5.22 to 1.5.23.

Updates `ch.qos.logback:logback-classic` from 1.5.22 to 1.5.23
- [Release notes](https://github.com/qos-ch/logback/releases)
- [Commits](https://github.com/qos-ch/logback/compare/v_1.5.22...v_1.5.23)

Updates `ch.qos.logback:logback-core` from 1.5.22 to 1.5.23
- [Release notes](https://github.com/qos-ch/logback/releases)
- [Commits](https://github.com/qos-ch/logback/compare/v_1.5.22...v_1.5.23)

---
updated-dependencies:
- dependency-name: ch.qos.logback:logback-classic
  dependency-version: 1.5.23
  dependency-type: direct:production
  update-type: version-update:semver-patch
- dependency-name: ch.qos.logback:logback-core
  dependency-version: 1.5.23
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump nl.jqno.equalsverifier:equalsverifier (#2412)

Bumps [nl.jqno.equalsverifier:equalsverifier](https://github.com/jqno/equalsverifier) from 4.2.5 to 4.3.
- [Release notes](https://github.com/jqno/equalsverifier/releases)
- [Changelog](https://github.com/jqno/equalsverifier/blob/main/CHANGELOG.md)
- [Commits](https://github.com/jqno/equalsverifier/compare/equalsverifier-4.2.5...equalsverifier-4.3)

---
updated-dependencies:
- dependency-name: nl.jqno.equalsverifier:equalsverifier
  dependency-version: '4.3'
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump com.github.eirslett:frontend-maven-plugin (#2411)

Bumps [com.github.eirslett:frontend-maven-plugin](https://github.com/eirslett/frontend-maven-plugin) from 1.15.4 to 2.0.0.
- [Changelog](https://github.com/eirslett/frontend-maven-plugin/blob/master/CHANGELOG.md)
- [Commits](https://github.com/eirslett/frontend-maven-plugin/compare/frontend-plugins-1.15.4...frontend-plugins-2.0.0)

---
updated-dependencies:
- dependency-name: com.github.eirslett:frontend-maven-plugin
  dependency-version: 2.0.0
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump org.json:json from 20250517 to 20251224 (#2410)

Bumps [org.json:json](https://github.com/douglascrockford/JSON-java) from 20250517 to 20251224.
- [Release notes](https://github.com/douglascrockford/JSON-java/releases)
- [Changelog](https://github.com/stleary/JSON-java/blob/master/docs/RELEASES.md)
- [Commits](https://github.com/douglascrockford/JSON-java/commits)

---
updated-dependencies:
- dependency-name: org.json:json
  dependency-version: '20251224'
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump org.junit:junit-bom from 6.0.1 to 6.0.2 (#2415)

Bumps [org.junit:junit-bom](https://github.com/junit-team/junit-framework) from 6.0.1 to 6.0.2.
- [Release notes](https://github.com/junit-team/junit-framework/releases)
- [Commits](https://github.com/junit-team/junit-framework/compare/r6.0.1...r6.0.2)

---
updated-dependencies:
- dependency-name: org.junit:junit-bom
  dependency-version: 6.0.2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump logback.version from 1.5.23 to 1.5.24 (#2414)

Bumps `logback.version` from 1.5.23 to 1.5.24.

Updates `ch.qos.logback:logback-classic` from 1.5.23 to 1.5.24
- [Release notes](https://github.com/qos-ch/logback/releases)
- [Commits](https://github.com/qos-ch/logback/compare/v_1.5.23...v_1.5.24)

Updates `ch.qos.logback:logback-core` from 1.5.23 to 1.5.24
- [Release notes](https://github.com/qos-ch/logback/releases)
- [Commits](https://github.com/qos-ch/logback/compare/v_1.5.23...v_1.5.24)

---
updated-dependencies:
- dependency-name: ch.qos.logback:logback-classic
  dependency-version: 1.5.24
  dependency-type: direct:production
  update-type: version-update:semver-patch
- dependency-name: ch.qos.logback:logback-core
  dependency-version: 1.5.24
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump org.apache:apache from 35 to 36 (#2413)

Bumps [org.apache:apache](https://github.com/apache/maven-apache-parent) from 35 to 36.
- [Release notes](https://github.com/apache/maven-apache-parent/releases)
- [Commits](https://github.com/apache/maven-apache-parent/commits)

---
updated-dependencies:
- dependency-name: org.apache:apache
  dependency-version: '36'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

chrisdutz committed:
feat: Added a new PlcCertificateAuthentication type to the API module.

hutcheb committed:
feat: Add github link to header

chrisdutz committed:
refactor: Refactored the ConnectionStateListener to allow reporting more types of events.

chrisdutz committed:
fix: Updated the prereqisiteCheck to also deal with new versions of libpcap.

dependabot[bot] committed:
build(deps): bump golang.org/x/text from 0.32.0 to 0.33.0 in /plc4go (#2417)

Bumps [golang.org/x/text](https://github.com/golang/text) from 0.32.0 to 0.33.0.
- [Release notes](https://github.com/golang/text/releases)
- [Commits](https://github.com/golang/text/compare/v0.32.0...v0.33.0)

---
updated-dependencies:
- dependency-name: golang.org/x/text
  dependency-version: 0.33.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

chrisdutz committed:
chore: Cleanup after merging in changes from develop

chrisdutz committed:
chore: Added updated "values", "fields" and "utlis" modules, which are needed by the code-generation.

chrisdutz committed:
chore: Added a new code-generation module based on JavaPoet and integrated that into the build of all drivers.

chrisdutz committed:
chore: Updated the mspecs for the new style of handling encoding, byteOrder etc.

chrisdutz committed:
fix: Resolved some build issues related to missing apache headers.

chrisdutz committed:
chore: Updated the utils and tools to handle the updated api

chrisdutz committed:
chore: Finished implementing support for referencing DataIo types in optional fields.

chrisdutz committed:
chore: Fine-tuned the generator for some of the mspecs I added back recently.

chrisdutz committed:
chore: Fine-tuned the mspecs, I recently added back to be compatible with the new code-generator

chrisdutz committed:
chore: Updated the code generated by the code-generator as proof that it works (all modules compile successfully)

chrisdutz committed:
feat: Added the config code in a separate spi module.

chrisdutz committed:
feat: Added the audit-log functionality.

chrisdutz committed:
chore: Update of the knx vendor ids.

chrisdutz committed:
feat: Added the API module for the new transports.

chrisdutz committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into refactor/spi3

# Conflicts:
#	.mvn/extensions.xml
#	plc4go/go.mod
#	plc4go/go.sum
#	plc4j/spi/src/main/java/org/apache/plc4x/java/spi/connection/DefaultNettyPlcConnection.java
#	plc4j/spi/src/test/java/org/apache/plc4x/java/spi/connection/SingleProtocolStackConfigurerTest.java
#	pom.xml

chrisdutz committed:
chore: Added missing Apache headers to the README.md

chrisdutz committed:
feat: Added a first version of a new TCP transport including comprehensive test-suite.

chrisdutz committed:
feat: Added a first version of a new TCP transport including comprehensive test-suite.

chrisdutz committed:
feat: Added a first version of a new Serial transport including comprehensive test-suite (sharable)

chrisdutz committed:
fix: Updated the code to not fail the build, if "socat" is not installed, which we need to run the tests.

chrisdutz committed:
chore: Fine tuned the tcp transport a bit more.

splatch commented:
@chrisdutz I see a change in generated code, but I can't see any changes around transports SPI. If I understand correctly main point of this pull request is to update plc4j-spi to not rely on netty to initialize client. Do you have a separate branch for that work? Frame representations we generated before and with above changes interact only with `(Reader/Writer)Buffer`, thus this PR does not touch netty part yet. I believe we still need other half of the changes.

chrisdutz commented:
Hmmm ... that's odd ... I think this PR has become so big GitHub can't show it anymore ... if you have a look at https://github.com/apache/plc4x/tree/refactor/spi3/plc4j/transports ... it's there.

chrisdutz commented:
<img width="1275" height="665" alt="image" src="https://github.com/user-attachments/assets/aab87cd3-fc03-454b-bc96-77fa51870b2b" />

splatch commented:
Yeah, github UI couldn't get more than 3000 files, transports fall apart from diff. Will have a look on them!

chrisdutz committed:
feat: Re-Implemented the pcap-replay transport.

chrisdutz committed:
feat: Re-Implemented the raw-socket transport.

chrisdutz committed:
feat: Re-Implemented the test transport.

sruehl commented:
Changes look good, awesome work, keep them coming :)

chrisdutz committed:
feat: Implemented a new TLS transport (using pre-shared-keys, self-signed-certificates and shared trust authority)

chrisdutz committed:
feat: Re-Implemented the can transports.

chrisdutz committed:
chore: Added some optimizations for parsing temporal values.

chrisdutz committed:
chore: Disabled the tools that will be replaced anyway later in the project

chrisdutz committed:
feat: Aded some additonal methods and default NullAuthentication to the API module.

chrisdutz committed:
test: Completely cleaned up and rewrote a lot of the old test-utils module to match recent changes.

chrisdutz committed:
chore: Updated the KNX and the BacNet vendor ids.

chrisdutz committed:
feat: Defined a "cotp" protocol (stripped down version of the S7 mspec) and implemented a COTP transport. (First complex transport that has partial protocol functionality as a layered trasport)

chrisdutz committed:
chore: Ported the logic of the old Modbus driver to the new SPI3 (Added a testsuite)

chrisdutz commented:
I've just pushed a first port of the Modbus driver to the new SPI3 ... please have a look and leave some feedback here. I'm now going to pause for a few days so you all have time to review and reply.

sruehl commented:
wow awesome work @chrisdutz. Going through the changes, CoPilot also found some interessting things, but one thing that jumped to my eye was the use of the package name `functions`. I would propose to name it `operations` to not confuse it with `functions` used in programming.

chrisdutz commented:
I just took the liberty to run the manual test benchmark in both the old and the new driver:

Old driver:
```
=== Test Summary ===
Single Item:
  Average Read Time:  13,80 ms (15 requests, 207 ms total)
  Average Write Time: 12,00 ms (15 requests, 180 ms total)
  Average Time:       12,90 ms
Multi Item (each 15 items):
  Average Read Time:  11,81 ms (100 requests, 1181 ms total)
  Average Write Time: 175,98 ms (100 requests, 17598 ms total)
  Average Time:       93,90 ms
======================
Test executed in 21570ms

```
New driver:
```
=== Test Summary ===
Protocol:  Modbus TCP (modbus-tcp)
Transport: TCP (tcp)
Single Item:
  Average Read Time:  14,60 ms (15 requests, 219 ms total)
  Average Write Time: 11,53 ms (15 requests, 173 ms total)
  Average Time:       13,07 ms
#Multi Item (each 15 items):
  Average Read Time:  16,47 ms (100 requests, 1647 ms total)
  Average Write Time: 174,59 ms (100 requests, 17459 ms total)
  Average Time:       95,53 ms
======================
Test executed in 19863ms

```
The new connection handling is quite a bit faster than the old Netty version. However, I see the new driver is in average 1ms/request slower ... will look into what I could do in order to improve that.

chrisdutz committed:
chore: Addressed some performance issues, that were causing the new Modbus driver to perform 4-6ms slower than the old one ... now it's a tiny bit faster than the old one.

chrisdutz commented:
Ok ... after the last commit, the new driver performs even a bit better than the old one:

```
=== Test Summary ===
Protocol:  Modbus TCP (modbus-tcp)
Transport: TCP (tcp)
Single Item:
  Average Read Time:  13,20 ms (15 requests, 198 ms total)
  Average Write Time: 11,80 ms (15 requests, 177 ms total)
  Average Time:       12,50 ms
#Multi Item (each 15 items):
  Average Read Time:  11,73 ms (100 requests, 1173 ms total)
  Average Write Time: 176,12 ms (100 requests, 17612 ms total)
  Average Time:       93,93 ms
======================
Test executed in 19532ms

```

chrisdutz committed:
chore: Renamed all transports artifactIds to use the plural form of "transport" to match the directory structure and the naming convention of other modules.

chrisdutz committed:
feat: Officially added support for ModbusTCP via TLS.

chrisdutz committed:
feat: Added real configuration types after finding out there's a formal default port for TLS secured Modbus TCP (802)

chrisdutz committed:
chore: Updated the KNX and the BacNet vendor ids.

chrisdutz committed:
test: Added Integration Tests that test all 3 Modbus variants with TCP, UDP and TLS transports.

chrisdutz committed:
test: Added DriverTestsuite tests for each supported types (both single item and multi-item)

chrisdutz committed:
chore: Updated the KNX and the BacNet vendor ids.

andvasp commented:
Hi @chrisdutz ,

I am doing some test on s7 to read some variables and it is not working. It does not even connect to the Plc. The same test is working using the code of develop branch.

To run the test i needed to change at [s7Driver](https://github.com/apache/plc4x/blob/refactor/spi3/plc4j/drivers/s7/src/main/resources/META-INF/services/org.apache.plc4x.java.api.PlcDriver) file like shown below because the class does not exist in the repo. Is it exist in your development environment?

org.apache.plc4x.java.s7.readwrite.S7Driver -> org.apache.plc4x.java.s7.S7Driver

Thanks,
Anderson.

chrisdutz commented:
> Hi @chrisdutz ,
> 
> I am doing some test on s7 to read some variables and it is not working. It does not even connect to the Plc. The same test is working using the code of develop branch.
> 
> To run the test i needed to change at [s7Driver](https://github.com/apache/plc4x/blob/refactor/spi3/plc4j/drivers/s7/src/main/resources/META-INF/services/org.apache.plc4x.java.api.PlcDriver) file like shown below because the class does not exist in the repo. Is it exist in your development environment?
> 
> org.apache.plc4x.java.s7.readwrite.S7Driver -> org.apache.plc4x.java.s7.S7Driver
> 
> Thanks, Anderson.

Did you check out the branch and build the entire project? I did move the locations of the driver classes.. But if you bike the full project it should work.

Chris

andvasp commented:
> Did you check out the branch and build the entire project? I did move the locations of the driver classes.. But if you bike the full project it should work.
> 
> Chris

----------

Yes, the entire project builds successfully. However, I am also testing other drivers and ran into an issue.

The ServiceLoader is throwing an exception because the class path listed in  [s7/META-INF/services](https://github.com/apache/plc4x/blob/refactor/spi3/plc4j/drivers/s7/src/main/resources/META-INF/services/org.apache.plc4x.java.api.PlcDriver) is incorrect. To fix it locally, I had to modify the mapping:

**From**: org.apache.plc4x.java.s7.**readwrite**.S7Driver (does not exist)
**To:** org.apache.plc4x.java.s7.~~readwrite~~.S7Driver

It seems either the META-INF/services file is pointing to the wrong package, or the readwrite.S7Driver class was accidentally left out of the repository.

Could you please check which one is correct? Once iI know I can continue my tests for s7 PLC.

Thanks,
Anderson.

chrisdutz commented:
Ok ... so I've finished porting all drivers and merged in all recent changes from develop ... I'll be merging this PR back to develop in a few days.

chrisdutz committed:
feat: Implement a new SPI (Shared code base for all drivers)

      o	Implement new Read-/WriteBuffers that work without third party dependencies
      o	Implement an updated code-generation framework
      o	Implement a new system for pluggable transports
      o	Implement the code for the concept of layered protocol drivers
      -	Update existing drivers to the new SPI

chrisdutz committed:
chore: Updated generated code.

sruehl commented:
TY for the PR, changes have been merged into main. Great work

andvasp commented:
> > Did you check out the branch and build the entire project? I did move the locations of the driver classes.. But if you bike the full project it should work.
> > Chris
> 
> Yes, the entire project builds successfully. However, I am also testing other drivers and ran into an issue.
> 
> The ServiceLoader is throwing an exception because the class path listed in [s7/META-INF/services](https://github.com/apache/plc4x/blob/refactor/spi3/plc4j/drivers/s7/src/main/resources/META-INF/services/org.apache.plc4x.java.api.PlcDriver) is incorrect. To fix it locally, I had to modify the mapping:
> 
> **From**: org.apache.plc4x.java.s7.**readwrite**.S7Driver (does not exist) **To:** org.apache.plc4x.java.s7.~readwrite~.S7Driver
> 
> It seems either the META-INF/services file is pointing to the wrong package, or the readwrite.S7Driver class was accidentally left out of the repository.
> 
> Could you please check which one is correct? Once iI know I can continue my tests for s7 PLC.
> 
> Thanks, Anderson.

Hi @chrisdutz ,

Did you have a chance to check my comment above? Maybe I am missing something.

Thank you,
Anderson.

chrisdutz commented:
Hi @andvasp ... 

I just had another look and you were right ... the services file was pointing to the wrong class ... but the thing that really puzzles me, is that I was running the tests in the tests manual package and they use the service-lookup and were working fine. Thanks for bringing this back up ... I'll fix it in develop.

Chris

andvasp commented:
Hi @chrisdutz,

I have completed the initial tests for this refactoring using the HelloPlc4xRead example from plc4x-extras. I compared the develop branch after the merge (commit e0950449) against the version before the merge (commit a495bbf9).

Here are the results:

OPC UA: Working on both versions.
EIP: Working on both versions.
AB-ETH: Works on the old version, but fails on the new one.
S7: Works on the old version, but fails on the new one.

I will investigating further.

Notes & Issues:
EIP Tags: I am having trouble reading certain tags in both versions, even though they can be read successfully from the OPC Server.
Debug Logs: I noticed that the old version displayed detailed debug-level logs (see below). This no longer seems to happen. Has this behavior changed?

```
╔═S7Message═════════════════════════════════════════════════════════════════════════════════════════════════════════╗
║╔═protocolId╗╔═messageType╗╔═reserved╗╔═tpduReference╗╔═parameterLength╗╔═payloadLength╗╔═S7MessageResponseData═══╗║
║║  0x32 50  ║║   0x03 3   ║║0x0000 0 ║║  0x000b 11   ║║    0x0002 2    ║║   0x0006 6   ║║╔═errorClass╗╔═errorCode╗║║
║╚═══════════╝╚════════════╝╚═════════╝╚══════════════╝╚════════════════╝╚══════════════╝║║  0x00 0   ║║  0x00 0  ║║║
║                                                                                        ║╚═══════════╝╚══════════╝║║
║                                                                                        ╚═════════════════════════╝║
║╔═parameter/S7Parameter═══════════════════════════════════╗                                                        ║
║║╔═parameterType╗╔═S7ParameterReadVarResponse/numItems═══╗║                                                        ║
║║║    0x04 4    ║║                0x01 1                 ║║                                                        ║
║║╚══════════════╝╚═══════════════════════════════════════╝║                                                        ║
║╚═════════════════════════════════════════════════════════╝                                                        ║
║╔═payload/S7Payload/S7PayloadReadVarResponse/items/S7VarPayloadDataItem═════════════════════════════╗              ║
║║╔═returnCode/DataTransportErrorCode════════════╗╔═transportSize/DataTransportSize════╗╔═dataLength╗║              ║
║║║                 0xff 255 OK                  ║║           0x05 5 INTEGER           ║║ 0x0010 16 ║║              ║
║║╚══════════════════════════════════════════════╝╚════════════════════════════════════╝╚═══════════╝║              ║
║║╔═data═══════════════════════════════════════╗                                                     ║              ║
║║║0|02 a8                         '..        '║                                                     ║              ║
║║╚════════════════════════════════════════════╝                                                     ║              ║
║╚═══════════════════════════════════════════════════════════════════════════════════════════════════╝              ║
╚═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝
```

Best regards,
Anderson

chrisdutz commented:
> AB-ETH: Works on the old version, but fails on the new one. S7: Works on the old version, but fails on the new one.

Could you please explain a bit more what exactly is not working and how it's failing? With only this bit of information I unfortunately can't help much ... and could you please open Issues for each driver that's failing? This makes tracking them down a lot simpler.

And regarding the output ... in the old SPI we pulled in all of the stuff in one big jar. That contained the normal output, the json output, the xml output and the box-based output (which you are referring to) ... the SPI3 I made a lot more modular. So I opted out for having a hard dependency on the box-based layout for this sort of thing and added the concept of the audit-log. The driver calls audit log output calls which default to a no-op thing unless you add a dependency to the audit-log-impl in your application and configure the audit log via your connection string. Then it's a lot more verbose.

I should probably add a documentation on the audit-log on our website.

andvasp commented:
> Could you please explain a bit more what exactly is not working and how it's failing? With only this bit of information I unfortunately can't help much ... and could you please open Issues for each driver that's failing? This makes tracking them down a lot simpler.

Hi @chrisdutz!

This was just a brief report to keep you in the loop. No need to worry about helping me just yet. I will investigate further and open a issue with more details if it turns out to be a real bug and not just my test.

Thanks!

IsmoLeszczynski commented:
This looks great! I'll be testing this next week with my team using multiple drivers on our test benches, perfectly timed for when I actually have time to mess around with new stuff!

chrisdutz commented:
Hehe ... I knew that and did extra shifts to allow you to do that ;-) 
(Just kidding ... I was doing extra shifts in order to have SPI3 audited by Mythos ... guess the Orang Utan and his gang prevented that :-/ )
